substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSections.java

```diff
@@ -0,0 +1,2344 @@
+            // uint32 length ............ in bytes (not counting these 4 bytes)
+            // uint16 dwarf_version ..... always 2
+            // uint32 info_offset ....... offset of compilation unit on debug_info
+            // uint8 address_size ....... always 8
+            // uint8 segment_desc_size .. ???
+            //
+            // i.e. 12 bytes followed by padding
+            // aligning up to 2 * address size
+            //
+            // uint8 pad[4]
+            //
+            // followed by N + 1 times
+            //
+            // uint64 lo ................ lo address of range
+            // uint64 length ............ number of bytes in range
+            //
+            // where N is the number of ranges belonging to the compilation unit
+            // and the last range contains two zeroes
+
+            for (ClassEntry classEntry : primaryClasses) {
+                pos += DW_AR_HEADER_SIZE;
+                // align to 2 * address size
+                pos += DW_AR_HEADER_PAD_SIZE;
+                pos += classEntry.getPrimaryEntries().size() * 2 * 8;
+                pos += 2 * 8;
+            }
+            byte[] buffer = new byte[pos];
+            super.setContent(buffer);
+        }
+
+        @Override
+        public byte[] getOrDecideContent(Map<Element, LayoutDecisionMap> alreadyDecided, byte[] contentHint) {
+            Element textElement = getElement().getOwner().elementForName(".text");
+            LayoutDecisionMap decisionMap = alreadyDecided.get(textElement);
+            if (decisionMap != null) {
+                Object valueObj = decisionMap.getDecidedValue(LayoutDecision.Kind.VADDR);
+                if (valueObj != null && valueObj instanceof Number) {
+                    // this may not be the final vaddr for the text segment
+                    // but it will be close enough to make debug easier
+                    // i.e. to within a 4k page or two
+                    debugTextBase = ((Number) valueObj).longValue();
+                }
+            }
+            return super.getOrDecideContent(alreadyDecided, contentHint);
+        }
+
+        @Override
+        public void writeContent() {
+            byte[] buffer = getContent();
+            int size = buffer.length;
+            int pos = 0;
+
+            checkDebug(pos);
+
+            debug("  [0x%08x] DEBUG_ARANGES\n", pos);
+            for (ClassEntry classEntry : primaryClasses) {
+                int lastpos = pos;
+                int length = DW_AR_HEADER_SIZE + DW_AR_HEADER_PAD_SIZE - 4;
+                int cuIndex = classEntry.getCUIndex();
+                LinkedList<PrimaryEntry> classPrimaryEntries = classEntry.getPrimaryEntries();
+                // add room for each entry into length count
+                length += classPrimaryEntries.size() * 2 * 8;
+                length += 2 * 8;
+                debug("  [0x%08x] %s CU %d length 0x%x\n", pos, classEntry.getFileName(), cuIndex, length);
+             
```

Comment:
That is the right place to switch from Path to String objects.

Reply:
Yes, uses of file names and paths are now converted at the point of use.